From c8df648fd0f8c3997c407efe743b7a67cb5e83dd Mon Sep 17 00:00:00 2001
From: Yu Kuai <yukuai3@huawei.com>
Date: Thu, 8 Dec 2022 10:42:42 +0800
Subject: [PATCH 0823/1398] [Huawei] blk-iocost: track whether iocg is still
 online

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188107

--------------------------------

blkcg_gq->online can't be used in iocost because it get cleared only after
all policies are offlined. This patch add a new field 'online' in iocg.

Signed-off-by: Yu Kuai <yukuai3@huawei.com>
Signed-off-by: Li Nan <linan122@huawei.com>
---
 block/blk-iocost.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/block/blk-iocost.c b/block/blk-iocost.c
index c9d74785f52e..368e825b200e 100644
--- a/block/blk-iocost.c
+++ b/block/blk-iocost.c
@@ -486,6 +486,7 @@ struct ioc_gq {
 	u32				inuse;
 
 	u32				last_inuse;
+	bool				online;
 	s64				saved_margin;
 
 	sector_t			cursor;		/* to detect randio */
@@ -2928,6 +2929,7 @@ static void ioc_pd_init(struct blkg_policy_data *pd)
 	ioc_now(ioc, &now);
 
 	iocg->ioc = ioc;
+	iocg->online = true;
 	atomic64_set(&iocg->vtime, now.vnow);
 	atomic64_set(&iocg->done_vtime, now.vnow);
 	atomic64_set(&iocg->active_period, atomic64_read(&ioc->cur_period));
@@ -2962,6 +2964,8 @@ static void ioc_pd_offline(struct blkg_policy_data *pd)
 	if (ioc) {
 		spin_lock_irqsave(&ioc->lock, flags);
 
+		iocg->online = false;
+
 		if (!list_empty(&iocg->active_list)) {
 			struct ioc_now now;
 
-- 
2.17.1

